### PR TITLE
Cache Parser configurers when running on a slave

### PR DESCRIPTION
**Pull request by jerrinot**
# Description
This changeset caches Parser Configurer(s) when running on a slave.

Details:
applyConfiguration() invokes RPC when running on slave.
-> parsing 1000s XML files results in 1000s of RPCs.

### Changelog entries

Proposed changelog entries:

* Slaves cache returned cache configurers

### CODE REVIEW

**Comment by jerrinot** [comment]
There is no JIRA issue because I don't have an account yet. I created the account today but I have not received a password yet.

**Comment by jerrinot** [comment]
I am not sure if eternal caching is desired. I will be happy to introduce some TTL is needed. 

I believe the Parser Configurers abstraction used for TestNG tests only.

**Comment by kuisathaverat** [comment]
this PR could resolve https://issues.jenkins-ci.org/browse/JENKINS-10234

**Comment by daniel-beck** [comment]
@kuisathaverat Seems unrelated, the problem there is more likely a too permissive ant pattern/too large workspace. Notably `applyConfiguration` isn't in the thread dump. Why do you think this is related?

**Comment by jerrinot** [comment]
@daniel-beck: I agree the JENKINS-10175 looks somewhat differently. This is the stacktrace I am observing during the lengthy result processing:
```
"pool-1-thread-40 for CLI connection to https://hazelcast-l337.ci.cloudbees.com / waiting for hudson.remoting.Channel@6dc17b83:CLI connection to https://hazelcast-l337.ci.cloudbees.com" #182 prio=5 os_prio=0 tid=0x00007f82a8010000 nid=0x9dd05 in Object.wait() [0x00007f8362200000] 
java.lang.Thread.State: TIMED_WAITING (on object monitor) 
at java.lang.Object.wait(Native Method) 
at hudson.remoting.Request.call(Request.java:147) 
- locked <0x00000005963e5128> (a hudson.remoting.UserRequest) 
at hudson.remoting.Channel.call(Channel.java:780) 
at hudson.util.io.ParserConfigurator.applyConfiguration(ParserConfigurator.java:81) 
at hudson.tasks.junit.SuiteResult.parse(SuiteResult.java:124) 
at hudson.tasks.junit.TestResult.parse(TestResult.java:301) 
at hudson.tasks.junit.TestResult.parsePossiblyEmpty(TestResult.java:244) 
at hudson.tasks.junit.TestResult.parse(TestResult.java:175) 
at hudson.tasks.junit.TestResult.parse(TestResult.java:154) 
at hudson.tasks.junit.TestResult.<init>(TestResult.java:126) 
at hudson.tasks.junit.JUnitParser$ParseResultCallable.invoke(JUnitParser.java:132) 
at hudson.tasks.junit.JUnitParser$ParseResultCallable.invoke(JUnitParser.java:107) 
at hudson.FilePath$FileCallableWrapper.call(FilePath.java:2731) 
at hudson.remoting.UserRequest.perform(UserRequest.java:120) 
at hudson.remoting.UserRequest.perform(UserRequest.java:48) 
at hudson.remoting.Request$2.run(Request.java:332) 
at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68) 
at java.util.concurrent.FutureTask.run(FutureTask.java:266) 
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) 
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) 
at java.lang.Thread.run(Thread.java:745) 
```

File: core/src/main/java/hudson/util/io/ParserConfigurator.java [hunk 1]
```diff
@@ -73,22 +74,37 @@ public void configure(SAXReader reader, Object context) {}
...
-                    private static final long serialVersionUID = -2178106894481500733L;
-
-                    public Collection<ParserConfigurator> call() throws IOException {
-                        return new ArrayList<ParserConfigurator>(all());
-                    }
-                });
+        if (Jenkins.getInstanceOrNull() == null) {
+            all = fromCacheOrRemote();
         } else
             all = all();
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/util/io/ParserConfigurator.java]: brackets for else would be also useful

File: core/src/main/java/hudson/util/io/ParserConfigurator.java [hunk 2]
```diff
@@ -73,22 +74,37 @@ public void configure(SAXReader reader, Object context) {}
...
+
         for (ParserConfigurator pc : all)
             pc.configure(reader,context);
     }
+
+    private static Collection<ParserConfigurator> fromCacheOrRemote() throws IOException, InterruptedException {
+        if (cache != null) {
+            return cache;
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/util/io/ParserConfigurator.java]: Such approach won't work reliably, e,g, if a new plugin with `ParserConfigurator` gets dynamically loaded. Without cache invalidation logic in may become error-prone

**Review by oleg-nenashev — COMMENTED** [review]
I would like to discuss the plugin Dynamic loading case before we go forward with such approach.

P.S: This also an example of a serializable extension point, which would generally prefer to avoid. Not related to this change

File: core/src/main/java/hudson/util/io/ParserConfigurator.java (hunk 2 above)
Comment (jerrinot) [inline core/src/main/java/hudson/util/io/ParserConfigurator.java]: what would you recommend to do? can slave subscribe to receive informations about plugins loaded on a master? or is a simple time-to-live enough?

File: core/src/main/java/hudson/util/io/ParserConfigurator.java [hunk 3]
```diff
@@ -55,6 +55,7 @@
  */
 public abstract class ParserConfigurator implements ExtensionPoint, Serializable {
     private static final long serialVersionUID = -2523542286453177108L;
+    private static volatile Collection<ParserConfigurator> cache;
```
Comment (jglick) [inline core/src/main/java/hudson/util/io/ParserConfigurator.java]: `@Extension`s should generally not be cached in a `static` field, lest the wrong objects carry over from one `Jenkins` session to the next. If I understand correctly, this field is only used when this class is loaded in an agent JVM, which would be specific to one session, so it is not a problem.

File: core/src/main/java/hudson/util/io/ParserConfigurator.java (hunk 2 above)
Comment (jglick) [inline core/src/main/java/hudson/util/io/ParserConfigurator.java]: You can use `ExtensionListListener` on the master side, but then how would you notify the agent to expire its cache? Could become quite tricky.

**Review by jglick — CHANGES_REQUESTED** [review]
I would suggest deprecating this whole extension point as misconceived from the start, and make the `junit` plugin apply `XMLEntityResolver` on its own rather than registering it globally—it is pointless for anyone else doing parsing.

**Comment by jglick** [comment]
The reported stack traces in JENKINS-10175 are unrelated to this optimization.

**Comment by jerrinot** [comment]
@oleg-nenashev, @jglick: thanks a lot for review. What something like [this](https://github.com/jenkinsci/junit-plugin/pull/65/) work better for you?

File: core/src/main/java/hudson/util/io/ParserConfigurator.java (hunk 2 above)
Comment (jerrinot) [inline core/src/main/java/hudson/util/io/ParserConfigurator.java]: @jglick: I don't really know Jenkins codebase so I thought there could be a subscribe mechanism. 

If there is no such thing then I guess proper invalidation is difficult. So I can either use a static TTL or completely [avoid using the extension mechanism](https://github.com/jenkinsci/junit-plugin/pull/65)

**Comment by kuisathaverat** [comment]
@daniel-beck I have supposed that the next step of ParserConfigurator should be the parsing process and depends on of when you take the thread dump you could find the stack trace that @jerrinot has posted, also the behavior it is pretty similar.

**Comment by kuisathaverat** [comment]
@jglick @oleg-nenashev on this https://github.com/jenkinsci/junit-plugin/pull/65/files makes the change to XMLEntityResolver for the JUnit Plugin, it is for sure better than a change on code

**Comment by batmat** [comment]
Closing, as per my last comment.

@jerrinot please feel free to comment here anytime once/if you feel like getting back to this, and thanks again for the work. We're all happy to reopen anytime it's needed. 

Thanks!

**Comment by jerrinot** [comment]
@batmat: this PR was actually superseded by https://github.com/jenkinsci/junit-plugin/pull/65/, I just forgot to close it. sorry, for that!

**Comment by batmat** [comment]
No worries, thanks again for your involvement!

**Pull request closed without merging**